pkg/kubelet/kuberuntime/logs/logs.go

```diff
@@ -277,6 +277,27 @@ func (w *logWriter) write(msg *logMessage, addPrefix bool) error {
 	return nil
 }
 
+type fileCloser struct {
+	oldFile *os.File
+}
+
+func (fc *fileCloser) Close() error {
+	klog.V(4).InfoS("Closing log file", "name", fc.oldFile.Name())
+	return fc.oldFile.Close()
+}
+
+// NewFileToClose close the old file, save the newF and close it next time.
+func (fc *fileCloser) NewFileToClose(newF *os.File) error {
+	if fc.oldFile != nil {
+		klog.V(4).InfoS("Closing log file", "name", fc.oldFile.Name())
+		if err := fc.oldFile.Close(); err != nil {
+			return err
+		}
+	}
+	fc.oldFile = newF
+	return nil
+}
+
```

Comment:
This is better, but still I think the code is hard to read.
I'd propose to return back to the original short bugfix PR and do a refactoring work in a separate PR. 
Does this make sense?

Reply:
Do you mean this commit? https://github.com/xyz-li/kubernetes/commit/2ff8b7b52ce3af7312c8a97707a9614e7847495c

Then we must open the new log file, and close the previous opened file, assign newF to the variable `f`.
https://github.com/xyz-li/kubernetes/commit/2ff8b7b52ce3af7312c8a97707a9614e7847495c?diff=split&w=0#diff-5c455ea60fb38862bb012a440688cb9f8319e5e5895ce5599038037689120f05R375

I haven't found a method to eliminate the assignment `f = newF`.

Reply:
> Then we must open the new log file, and close the previous opened file, assign newF to the variable f.
https://github.com/xyz-li/kubernetes/commit/059d2c5abedb60ba0ae2d5c6f23683f2c1d4197d#diff-5c455ea60fb38862bb012a440688cb9f8319e5e5895ce5599038037689120f05R376-R377

yes, please do. That should fix the issue.  
And then, if you want to continue, you can refactor the code in a separate PR.

Reply:
#124298. This is actually the initial commit.

Reply:
Thanks. I'll lgtm it if/when CI is green.